app/com/linkedin/drelephant/ElephantRunner.java

```diff
@@ -45,136 +48,377 @@
               logger.info("Error with hadoop kerberos login", e);
-              //Wait for a while before retry
-              waitInterval();
-              continue;
-            }
-
-            List<AnalyticJob> todos;
-            try {
-              todos = _analyticJobGenerator.fetchAnalyticJobs();
-            } catch (Exception e) {
-              logger.error("Error fetching job list. Try again later...", e);
-              //Wait for a while before retry
-              waitInterval();
+              waitInterval(_fetchInterval);
               continue;
             }
 
-            _jobQueue.addAll(todos);
-            logger.info("Job queue size is " + _jobQueue.size());
-
-            //Wait for a while before next fetch
-            waitInterval();
+            fetchApplications(_lastTime + 1, _currentTime);
+            waitInterval(_fetchInterval);
           }
           logger.info("Main thread is terminated.");
           return null;
         }
       });
-    } catch (Exception e) {
+    } catch (IOException e) {
       logger.error(e.getMessage());
       logger.error(ExceptionUtils.getStackTrace(e));
+      _completedJobPool.shutdown();
+      _undefinedJobPool.shutdown();
     }
   }
 
-  private class ExecutorThread implements Runnable {
+  private void initialize() {
+    if (!HadoopSystemContext.isHadoop2Env()) {
+      throw new RuntimeException("Unsupported Hadoop major version detected. It is not 2.x.");
+    }
+
+    try {
+      _analyticJobGenerator.configure(_configuration);
+    } catch (Exception e) {
+      logger.error("Error occurred when configuring the analysis provider.", e);
+      throw new RuntimeException(e);
+    }
+
+    HDFSContext.load();
+    ElephantContext.init();
+    loadGeneralConfiguration();
+    loadExecutorThreads();
+  }
+
+  /**
+   * Load all the properties from GeneralConf.xml
+   */
+  private void loadGeneralConfiguration() {
+    _configuration.addResource(this.getClass().getClassLoader().getResourceAsStream("GeneralConf.xml"));
+
+    _completedExecutorCount = getLongFromConf(_configuration, COMPLETED_JOB_POOL_SIZE_KEY, DEFAULT_COMPLETED_JOB_POOL_SIZE);
+    _runningExecutorCount = getLongFromConf(_configuration, RUNNING_JOB_POOL_SIZE_KEY, DEFAULT_RUNNING_JOB_POOL_SIZE);
+
+    _fetchInterval = getLongFromConf(_configuration, FETCH_INTERVAL_KEY, DEFAULT_FETCH_INTERVAL);
+    _fetchLag = getLongFromConf(_configuration, FETCH_LAG_KEY, DEFAULT_FETCH_LAG);
+
+    _runningJobUpdateInterval = getLongFromConf(_configuration, RUNNING_JOB_UPDATE_INTERVAL_KEY,
+        DEFAULT_RUNNING_JOB_UPDATE_INTERVAL);
+  }
+
+  /**
+   * Extract a long value from the configuration
+   *
+   * @param conf The configuration object
+   * @param key The key to be extracted
+   * @param defaultValue The default value to use when key is not found
+   * @return the extracted value
+   */
+  private long getLongFromConf(Configuration conf, String key, long defaultValue) {
```

Comment:
Well, I did think about moving it to a utility class. But if you carefully take a look, the exception contains a message specific to GeneralConf.xml. Now, we can pass the conf file name as a string, but then the class will have too many arguments. So I left it in this class itself. And no other class is currently using it.